neo/Network/neonetwork/network/protocol.py

```diff
@@ -0,0 +1,95 @@
+import asyncio
+import struct
+from typing import Optional
+from neo.Network.neonetwork.network.node import NeoNode
+from neo.Network.neonetwork.network.message import Message
+from asyncio.streams import StreamReader, StreamReaderProtocol, StreamWriter
+from asyncio import events
+
+
+class NeoProtocol(StreamReaderProtocol):
+    def __init__(self, *args, quality_check=False, **kwargs):
+        """
+
+        Args:
+            *args:
+            quality_check (bool): there are times when we only establish a connection to check the quality of the node/address
+            **kwargs:
+        """
+        self._stream_reader = StreamReader()
+        self._stream_writer = None
+        nodemanager = kwargs.pop('nodemanager')
+        self.client = NeoNode(self, nodemanager, quality_check)
+        self._loop = events.get_event_loop()
+        super().__init__(self._stream_reader)
+
+    def connection_made(self, transport: asyncio.transports.BaseTransport) -> None:
+        super().connection_made(transport)
+        self._stream_writer = StreamWriter(transport, self, self._stream_reader, self._loop)
+
+        if self.client:
+            asyncio.create_task(self.client.connection_made(transport))
+
+    def connection_lost(self, exc: Optional[Exception] = None) -> None:
+        super().connection_lost(exc)
+        if self.client:
+            self.client.connection_lost(exc)
+
+    def eof_received(self) -> bool:
+        self._stream_reader.feed_eof()
+
+        self.connection_lost()
+        return True
+        # False == Do not keep connection open, this makes sure that `connection_lost` gets called.
+        # return False
+
+    async def send_message(self, message: Message) -> None:
+        try:
+            self._stream_writer.write(message.to_array())
+            await self._stream_writer.drain()
+        except ConnectionResetError:
+            print("connection reset")
+            self.connection_lost(ConnectionResetError)
+            self.disconnect()
+        except ConnectionError:
+            print("connection error")
+            self.connection_lost(ConnectionError)
+            self.disconnect()
+        except asyncio.CancelledError:
+            print("task cancelled, closing connection")
```

Comment:
I think these should be debug statements

Reply:
regarding these print statements (and those below) . Yes, eventually they shouldn't be there at all. I think even including them with `logger.debug` might generate too much noise. I'll removing them as a TODO but want to keep them for now because I think they're useful at this stage